instrumentation/opentelemetry_grpc/lib/opentelemetry_grpc/server.ex

```diff
@@ -0,0 +1,266 @@
+  defp build_server_attributes(metadata) do
+    %{
+      RPCAttributes.rpc_system() => @grpc_rpc_system,
+      RPCAttributes.rpc_service() => metadata.stream.service_name,
+      RPCAttributes.rpc_method() => metadata.stream.method_name,
+      NetworkAttributes.network_protocol_name() => "http",
+      NetworkAttributes.network_protocol_version() => "2",
+      NetworkAttributes.network_transport() => "tcp",
+      "elixir.grpc.server_module" => to_string(metadata.server)
+    }
+    |> put_endpoint_module_attr(metadata.endpoint)
+  end
+
+  defp put_endpoint_module_attr(attrs, nil) do
+    attrs
+  end
+
+  defp put_endpoint_module_attr(attrs, endpoint) do
+    Map.put(attrs, "elixir.grpc.endpoint_module", to_string(endpoint))
+  end
+
+  defp record_server_span_result({:ok, _stream, _response}) do
+    Tracer.set_attributes(%{
+      RPCAttributes.rpc_grpc_status_code() => @status_code_ok
+    })
+
+    Tracer.set_status(OpenTelemetry.status(:ok))
+  end
+
+  defp record_server_span_result({:error, %GRPC.RPCError{status: status, message: message}}) do
+    Tracer.set_attributes(%{
+      RPCAttributes.rpc_grpc_status_code() => Map.get(@status_code_mapping, status, status),
+      ErrorAttributes.error_type() => to_string(GRPC.RPCError)
+    })
+
+    status_name = GRPC.Status.code_name(status)
+    Tracer.set_status(OpenTelemetry.status(:error, message || "gRPC error: #{status_name}"))
+  end
+
+  defp record_server_span_result({:error, %error_struct{} = error}) when is_exception(error) do
+    Tracer.set_attributes(%{
+      RPCAttributes.rpc_grpc_status_code() => @status_code_internal,
+      ErrorAttributes.error_type() => to_string(error_struct)
+    })
+
+    Tracer.set_status(OpenTelemetry.status(:error, Exception.message(error)))
+  end
+
+  defp record_server_span_result(_result) do
+    # Unknown result format - don't set status
+    :ok
+  end
+
+  defp put_links(span_opts, []) do
+    span_opts
+  end
+
+  defp put_links(span_opts, links) do
+    Map.put(span_opts, :links, links)
+  end
+
+  defp setup_context_propagation(source, :child) do
+    extract_and_attach(source)
+  end
+
+  defp setup_context_propagation(source, :link) do
+    link_from_propagated_ctx(source)
+  end
+
+  defp setup_context_propagation(_source, _) do
+    []
+  end
+
+  defp extract_and_attach(source) do
+    case get_propagated_ctx(source) do
+      {_links, parent_ctx} when parent_ctx != :undefined ->
+        OpenTelemetry.Ctx.attach(parent_ctx)
+
+        # When we attach the context, we don't need links - parent-child relationship is established
+        []
+
+      {links, _undefined_ctx} ->
+        # No parent context to attach, but we can still return links if any
+        links
+    end
+  end
+
+  defp link_from_propagated_ctx(source) do
+    {links, _ctx} = get_propagated_ctx(source)
+    links
+  end
+
+  defp get_propagated_ctx(stream) do
+    stream
+    |> get_grpc_headers()
+    |> extract_to_ctx()
+  end
+
+  defp extract_to_ctx([]) do
```

Comment:
Will an empty array or the presence of items always be returned? Any case clause risk?

Reply:
let me follow up on this tomorrow, maybe I shouldnt short-curcuit here, now thinking if we still want to do the linking stuff which now I am not doing 😓